test/api/model/events_test.dart

```diff
@@ -42,4 +29,35 @@ void main() {
         '      on the pattern of the existing cases.'
     ).isEmpty();
   });
+
+  test('message: move flags into message object', () {
+    final message = eg.streamMessage();
+    MessageEvent mkEvent(List<MessageFlag> flags) => Event.fromJson({
+      'type': 'message',
+      'id': 1,
+      'message': (deepToJson(message) as Map<String, dynamic>)..remove('flags'),
+      'flags': flags.map((f) => f.toJson()).toList(),
+    }) as MessageEvent;
+    check(mkEvent(message.flags)).message.jsonEquals(message);
+    check(mkEvent([])).message.flags.deepEquals([]);
+    check(mkEvent([MessageFlag.read])).message.flags.deepEquals([MessageFlag.read]);
+  });
+
+  test('update_message_flags/remove: require messageDetails in mark-as-unread', () {
+    final baseJson = {
+      'id': 1,
+      'type': 'update_message_flags',
+      'op': 'remove',
+      'flag': 'starred',
+      'messages': [123],
+      'all': false,
+    };
+    check(() => UpdateMessageFlagsRemoveEvent.fromJson(baseJson)).returnsNormally();
+    check(() => UpdateMessageFlagsRemoveEvent.fromJson({
+      ...baseJson, 'flag': 'read',
+    })).throws();
+    check(() => UpdateMessageFlagsRemoveEvent.fromJson({
+      ...baseJson, 'message_details': {'123': {'type': 'private', 'mentioned': false, 'user_ids': [2]}},
+    })).returnsNormally();
```

Comment:
This test should set `'flag': 'read'` too, right?

Reply:
Eep, thanks! 89fd1f79a